### ISSUE TITLE
deeplabcut.create_training_dataset: "FileNotFoundError: No such file:"

### ISSUE THREAD

**Issue by ninaharano** [DeepLabCut/DeepLabCut#928]
Some similar questions regarding the step have been asked in the platform, but since none could solve my issue I was wondering if I can receive any feedback.

I run into FileNotFound error upon running the line "deeplabcut.create_training_dataset(config_path)."

(1) The path should not be wrong because I can run "deeplabcut.check_labels(conﬁg_path)" with no problem. 

(2) In addition, the file that is said to be non-existent in the error does indeed exist within labeled-data folder. 

I would appreciate any small hints. Thank you so much in advance!

BELOW IS THE CODE 
```
In [8]: deeplabcut.create_training_dataset(config_path)                         
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-8-c980124be66b> in <module>
----> 1 deeplabcut.create_training_dataset(config_path)

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/deeplabcut/generate_training_dataset/trainingsetmanipulation.py in create_training_dataset(config, num_shuffles, Shuffles, windows2linux, userfeedback, trainIndices, testIndices, net_type, augmenter_type)
   1007                 ################################################################################
   1008                 data, MatlabData = format_training_data(
-> 1009                     Data, trainIndices, nbodyparts, project_path
   1010                 )
   1011                 sio.savemat(

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/deeplabcut/generate_training_dataset/trainingsetmanipulation.py in format_training_data(df, train_inds, nbodyparts, project_path)
    788         filename = df.index[i]
    789         data["image"] = filename
--> 790         img_shape = _read_image_shape_fast(os.path.join(project_path, filename))
    791         try:
    792             data["size"] = img_shape[2], img_shape[0], img_shape[1]

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/deeplabcut/generate_training_dataset/trainingsetmanipulation.py in _read_image_shape_fast(path)
    772 @lru_cache(maxsize=None)
    773 def _read_image_shape_fast(path):
--> 774     return io.imread(path).shape
    775 
    776 

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/skimage/io/_io.py in imread(fname, as_gray, plugin, **plugin_args)
     46 
     47     with file_or_url_context(fname) as fname:
---> 48         img = call_plugin('imread', fname, plugin=plugin, **plugin_args)
     49 
     50     if not hasattr(img, 'ndim'):

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/skimage/io/manageplugins.py in call_plugin(kind, *args, **kwargs)
    207                                (plugin, kind))
    208 
--> 209     return func(*args, **kwargs)
    210 
    211 

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/skimage/io/_plugis/imageio_plugin.py in imread(*args, **kwargs)
      8 @wraps(imageio_imread)
      9 def imread(*args, **kwargs):
---> 10     return np.asarray(imageio_imread(*args, **kwargs))

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/imageio/core/funcions.py in imread(uri, format, **kwargs)
    263 
    264     # Get reader and read first
--> 265     reader = read(uri, format, "i", **kwargs)
    266     with reader:
    267         return reader.get_data(0)

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/imageio/core/funcions.py in get_reader(uri, format, mode, **kwargs)
    170 
    171     # Create request object
--> 172     request = Request(uri, "r" + mode, **kwargs)
    173 
    174     # Get format

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/imageio/core/requst.py in __init__(self, uri, mode, **kwargs)
    119 
    120         # Parse what was given
--> 121         self._parse_uri(uri)
    122 
    123         # Set extension

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/imageio/core/requst.py in _parse_uri(self, uri)
    255                 # Reading: check that the file exists (but is allowed adir)
    256                 if not os.path.exists(fn):
--> 257                     raise FileNotFoundError("No such file: '%s'" % fn)
    258             else:
    259                 # Writing: check that the directory to write to does exst

FileNotFoundError: No such file: '/labeled-data/○○○/img082532.png'
```

**Comment by AlexEMG** [comment]
This particular image seems to be missing, so you cannot create the training set. You should first apply check_labels and see if you annotated everything and if the data is complete. If it is not, you should remove images that have no labels, or vice versa annotations to images that do not exist. 

Check out the [helper functions for this](https://github.com/DeepLabCut/DeepLabCut/blob/master/docs/HelperFunctions.md), in particular:

https://github.com/DeepLabCut/DeepLabCut/blob/b1f8f413ec8f5ffdb522624b62a02b5980472f66/deeplabcut/generate_training_dataset/trainingsetmanipulation.py#L160

**Comment by ninaharano** [comment]
**Thank you so much for your feedback.
Since I am still lost, I would appreciate it if you would provide a little more additional help...!**

(1) I manually checked that **the particular image that is said to be missing was in the** "/.../labeled-data/projectname_labeled" **folder**, and every time I repeat the deeplabcut.create_training_dataset(config_path) line, it turns out with the same error but with different image name. **I saw that the images in the folder are named "img082532_bodypart.png" instead of "img082532.png." Could it be the source of error?**

(2) **All the images seem to have labels (dots) on the picture when I manually check them in the file.** 

**Running the check_labels command** turns out the following:

```
> In [5]: deeplabcut.check_labels(config_path)                                    
> Creating images with labels by ....
> /mnt/expanse/homes/.../labeled-data/..._labeled  already exists!
> 100%|âââââââââââââââââââââââââââââââââââââââââââ| 29/29 [00:34<00:00,  1.18s/it]
> If all the labels are ok, then use the function 'create_training_dataset' to create the training dataset!
> 
```

Running the dropannotationfileentriesduetodeletedimages code seems to drop annotation for all of my images (not sure if this is how it's supposed to be).

```

Dropping... /labeled-data/JP56 20190411/img004292.png
Dropping... /labeled-data/JP56 20190411/img012169.png
Dropping... /labeled-data/JP56 20190411/img022680.png
Dropping... /labeled-data/JP56 20190411/img024958.png
Dropping... /labeled-data/JP56 20190411/img031985.png
Dropping... /labeled-data/JP56 20190411/img052370.png
Dropping... /labeled-data/JP56 20190411/img061391.png
Dropping... /labeled-data/JP56 20190411/img071573.png
Dropping... /labeled-data/JP56 20190411/img082532.png
Dropping... /labeled-data/JP56 20190411/img094080.png
Dropping... /labeled-data/JP56 20190411/img094326.png
Dropping... /labeled-data/JP56 20190411/img101284.png
Dropping... /labeled-data/JP56 20190411/img115389.png
Dropping... /labeled-data/JP56 20190411/img133471.png
Dropping... /labeled-data/JP56 20190411/img134686.png
Dropping... /labeled-data/JP56 20190411/img143734.png
Dropping... /labeled-data/JP56 20190411/img185194.png
Dropping... /labeled-data/JP56 20190411/img188906.png
Dropping... /labeled-data/JP56 20190411/img189517.png
Dropping... /labeled-data/JP56 20190411/img189596.png
Dropping... /labeled-data/JP56 20190411/img199304.png
Dropping... /labeled-data/JP56 20190411/img212699.png
Dropping... /labeled-data/JP56 20190411/img223130.png
Dropping... /labeled-data/JP56 20190411/img223262.png
Dropping... /labeled-data/JP56 20190411/img231118.png
Dropping... /labeled-data/JP56 20190411/img258787.png
Dropping... /labeled-data/JP56 20190411/img260389.png
Dropping... /labeled-data/JP56 20190411/img276594.png
Dropping... /labeled-data/JP56 20190411/img289945.png


```

**Running the create_training_dataset line after running the above line turns out the following.**

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-11-c980124be66b> in <module>
----> 1 deeplabcut.create_training_dataset(config_path)

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/deeplabcut/generate_training_dataset/trainingsetmanipulation.py in create_training_dataset(config, num_shuffles, Shuffles, windows2linux, userfeedback, trainIndices, testIndices, net_type, augmenter_type)
    899 
    900         Data = merge_annotateddatasets(
--> 901             cfg, Path(os.path.join(project_path, trainingsetfolder)), windows2linux
    902         )
    903         if Data is None:

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/deeplabcut/generate_training_dataset/trainingsetmanipulation.py in merge_annotateddatasets(cfg, trainingsetfolder_full, windows2linux)
    635     # Let's check if the code is *not* run on windows (Source: #https://stackoverflow.com/questions/1325581/how-do-i-check-if-im-running-on-windows-in-python)
    636     # but the paths are in windows format...
--> 637     windowspath = "\\" in AnnotationData.index[0]
    638     if os.name != "nt" and windowspath and not windows2linux:
    639         print(

/usr/local/bin/anaconda3/envs/DLC/lib/python3.7/site-packages/pandas/core/indexes/base.py in __getitem__(self, key)
   3928         if is_scalar(key):
   3929             key = com.cast_scalar_indexer(key)
-> 3930             return getitem(key)
   3931 
   3932         if isinstance(key, slice):

IndexError: index 0 is out of bounds for axis 0 with size
``` 

**If you could bear with me and provide additional feedbacks, that would help me so much. Thank you!**

**Comment by MMathisLab** [comment]
>>  I saw that the images in the folder are named "img082532_bodypart.png" instead of "img082532.png." Could it be the source of error? 


yes that is incorrect; do you know how that happened? 


>> dropannotationfileentriesduetodeletedimages 

this means now all your annotations are dropped :(... I hope you have a backup!

**Comment by jeylau** [comment]
You are looking at the wrong folder: labeled-data/videoname_labeled rightly contains images 'labeled' after running check_labels. The original images are in labeled-data/videoname.
What is your project path like (the one you find in config.yaml)?

**Comment by ninaharano** [comment]
Thank you so much for the feedback.

@MMathisLab I am not sure how the wrong folder names were created. 
I don't have annotations backed up, so I will go re-do some steps. Since all the annotations are dropped, I am assuming that I should re-run all the steps after frame extraction.?

@jeylau The project path I found in my config.yaml file is "_project_path: /mnt/expanse/homes/my_name/.../project_name/project_name-my_name-2020-09-17_" - I was not awareany chance this is wrong? This + config.yaml at the end is what I use for deeplabcut.check_labels(config_path) command, which does not work for me.

And yes! I do have two folders (labeled-data/videoname_labeled and labeled-data/videoname) in my file. I can see the pictures with labels in the "labeled-data/videoname_labeled" folder. If this seems problematic, I would appreciate it if you would briefly let me know what it could/should look like. 

Thank you so much for all your generous help. I am stuck but if I could make it run I would appreciate it so much..!

**Comment by MMathisLab** [comment]
` labeled-data/videoname` <-- only the raw images in this are used for training, so if they are gone, that is the issue! No clue how they would get deleted though! And, that would explain it all...

**Comment by GPU-Shuttle** [comment]
Here's the somehow similar error I am facing atm!
I had a project setup and working on the server; I was wanting to add a new participant to the dataset; hence labelled data from that new participant.
Then I created a new project on the server by including video from the new participant in the initial set-up. Then copied the previous "labeled-data" folder to the new project and also uploaded/added the labeled data (including the .csv and .h5 files) from my pc to the "labeled-data";
I face the 
".......raise FileNotFoundError("No such file: '%s'" % fn)
FileNotFoundError: No such file:" 
error once I want to create the training dataset! 
The "deeplabcut.dropannotationfileentriesduetodeletedimages(config)" command drops all the images related to the subject that I have labeled on my pc (and removes the markers coordinates from the .csv file :-( )! not sure what I should do to solve this. I would appreciate if you could comment on that, please?

**Comment by MMathisLab** [comment]
^ @GPU-Shuttle to add more videos, you should do so in the same project - you simply can either (1) run DLC on this new participant, then extract outliers, refine and MERGE. This creates a new iteration and your original models are fully intact.  This is 100% what I recommend and what is described in our Protocol paper : Nath et al 2019 Nature Protocols. (2) run add_new_videos, extract frames, label, and then create a NEW SHUFFLE #, such that is does not overwrite your original work.

**Comment by GPU-Shuttle** [comment]
Thanks, I feel like I know what the problem is, but I don't know how I'd be able to fix it!
It's definitely related to path issues.
I realized that all the paths to the images in the .csv file from the previous dataset in the 'labeled-data' folder on the server have "/" to address the path  (e.g., labeled-data/P08_ML/img0002.png); whereas the .csv file for the new participant shows " \ " instead (e.g., labeled-data\P10_ML\img0018.png). I think all the " \ "s need to be changed to "/"s. Now wonder how'd be possible to edit this in the .h5 file as I am sure DLC reads from that and not from the .csv file? 

Thanks again

**Comment by GPU-Shuttle** [comment]
This is now solved under issue #1134.
Thanks
